sound/soc/sof/ipc4-pcm.c

```diff
@@ -571,11 +576,67 @@ static int sof_ipc4_pcm_get_time_info(struct snd_soc_component *component,
 	return 0;
 }
 
+static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_llp_reading_slot llp, last_llp;
+	struct snd_sof_pcm_stream *stream;
+	struct snd_sof_pcm *spcm;
+	snd_pcm_uframes_t pos;
+	int time = 100;
+	u64 link_pos;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
+
+	stream = &spcm->stream[substream->stream];
+
+	if (stream->llp_offset) {
+		/*
+		 * Host dma position is gotten from mmio in host space but llp counter
+		 * is updated by fw by each schedule time, 1ms in common case. To
+		 * synchronize them, host will read host dma position when the llp counter
+		 * is just changed by fw. Driver checks the llp change every 10 us, so the
+		 * error time will about 10 ~ 15 us.
```

Comment:
I'm not 100% comfortable with the usleep loop. Do we really need accuracy at this level? When an apps uses deep buffering, some inaccuracy is to be expected do to external buffering.

Reply:
@plbossart  without loop method, the max error time will be 1ms, and average error is 0.5ms. Do we need loop method to get high accuracy for delay function ? It is much simple with loop

Reply:
We are also looping here for worst case 1ms, iow, as long as the llp_offset changes.
If we happen to read the last_llp right after it got changed we will wait for the next update.

The llp_offset is updated in what interval?

Reply:
Thanks @RanderWang for explanation! I think any sleep on the delay() path is problematic (from app developer perspective), so I'd prefer to go without the loop method and accept the inaccuracy.

Reply:
sure